### ISSUE TITLE
Analyzer rule for using the "is" operator on unrelated types

### ISSUE THREAD

**Issue by noinskit** [dart-lang/linter#2450]
I'd love having a compiler warning and/or analyzer rule to warn about code like:
```
main() async {
  print("${5 is String}");
}
```

This code is correct (and the answer is "false"), but it's already known at compile time, from static analysis, and I believe in practice just signals a bug. Basically, when I use the "is" operator I actually mean "is subtype".

This would be similar to the already existing analyzer check `unrelated_type_equality_checks`.

This actually bit me during refactoring in a production project where the type of a method changed and one client of the method was not updated, leading to suboptimal error handling on user devices.

**Comment by srawlins** [comment]
`is` is also a valid operator given an unrelated type. For example:

```dart
class C {}
class D {}
class E implements C, D {}

f(C c) {
  if (c is D) {
    // Totally valid, as there can be classes which implement both C and D, such as E.
  }
}
```

I can still see this being a useful check, but because of the false positive possibilities, I'm moving this request over to the linter.

**Comment by cedvdb** [comment]
How would it work though ?

as per @srawlins comment's it has valid uses

**Comment by bwilkerson** [comment]
There isn't any way to statically determine whether such an expression is intentional, and it would require whole-program analysis (which the linter can't do) to determine whether or not it is possible for such an expression to evaluate to `true`. My guess, then, is that the lint would be overly restrictive, flagging places like the code that @srawlins showed, and that users would then be required to place an `ignore` comment in those few places where they needed to use such a code pattern.

Before we implement such a lint, I think it would be worth doing a quick investigation to try to figure out what the false positive rate for such a lint would be in practice.
